internal/pipe/srpm/srpm.go

```diff
@@ -0,0 +1,229 @@
+// Package srpm implements the Pipe interface building source RPMs.
+package srpm
+
+import (
+	_ "embed"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/caarlos0/log"
+	"github.com/goreleaser/goreleaser/internal/artifact"
+	"github.com/goreleaser/goreleaser/internal/skips"
+	"github.com/goreleaser/goreleaser/internal/tmpl"
+	"github.com/goreleaser/goreleaser/pkg/context"
+	"github.com/goreleaser/nfpm/v2"
+	"github.com/goreleaser/nfpm/v2/files"
+
+	_ "github.com/goreleaser/nfpm/v2/rpm" // blank import to register the format
+)
+
+var (
+	defaultFileNameTemplate     = "{{ .PackageName }}-{{ .Version }}.src.rpm"
+	defaultSpecFileNameTemplate = "{{ .PackageName }}.spec"
+)
+
+//go:embed spec.tmpl
+var defaultSpecTemplate string
+
+// Pipe for source RPMs.
+type Pipe struct{}
+
+func (Pipe) String() string { return "source RPMs" }
+func (Pipe) Skip(ctx *context.Context) bool {
+	return skips.Any(ctx, skips.SRPM) || !ctx.Config.SRPM.Enabled
+}
+
+// Default sets the pipe defaults.
+func (Pipe) Default(ctx *context.Context) error {
+	srpm := &ctx.Config.SRPM
+	if srpm.ID == "" {
+		srpm.ID = "default"
+	}
+	if srpm.PackageName == "" {
+		srpm.PackageName = ctx.Config.ProjectName
+	}
+	if srpm.FileNameTemplate == "" {
+		srpm.FileNameTemplate = defaultFileNameTemplate
+	}
+	if srpm.SpecFileNameTemplate == "" {
+		srpm.SpecFileNameTemplate = defaultSpecFileNameTemplate
+	}
+	if srpm.SpecTemplate == "" {
+		srpm.SpecTemplate = defaultSpecTemplate
+	}
+	if srpm.Bins == nil {
+		srpm.Bins = map[string]string{
+			ctx.Config.ProjectName: "%{goipath}",
+		}
+	}
+	return nil
+}
+
+// Run the pipe.
+func (Pipe) Run(ctx *context.Context) error {
+	sourceArchives := ctx.Artifacts.Filter(artifact.ByType(artifact.UploadableSourceArchive)).List()
```

Comment:
Seems the result for me here is always `nil`. But I expect my test config might be wrong?

May I ask, do you have a config that you used for testing? :) 

<details>

<summary>Here is what I used so far</summary>

```yaml
---
version: 1

env:
  - GO111MODULE=on

before:
  hooks:
    - go mod tidy

gomod:
  proxy: true

report_sizes: true

git:
  ignore_tags:
    - "{{ if not .IsNightly }}nightly{{ end }}"

metadata:
  mod_timestamp: "{{ .CommitTimestamp }}"

builds:
  - env:
      - CGO_ENABLED=0
    goos:
      - linux
    goarch:
      - amd64
    mod_timestamp: "{{ .CommitTimestamp }}"
    main: ./cmd/vlookup/main.go
    flags:
      - -trimpath
    ldflags:
      - -s -w -X main.version={{.Version}} -X main.commit={{.Commit}} -X main.date={{ .CommitDate }} -X main.builtBy=goreleaser -X main.treeState={{ .IsGitDirty }}

universal_binaries:
  - replace: false

checksum:
  name_template: "checksums.txt"

srpm:
```

</details>

Reply:
I've not run the code in this PR for a long time. I think you need to build a source archive, something like:

```yaml
source:
  enabled: true
  prefix_template: '{{ .ProjectName }}-{{ .Version }}/'
```